### PR TITLE
[CDF-24555] 📂 Support dump files

### PR DESCRIPTION
**Pull request by doctrino**
# Description

Adds support for dumping file metadata to csv/parquet.

image

## Changelog

- [x] Patch
- [ ] Minor
- [ ] Skip

### Added

- [alpha] Support for dumping files medata, `cdf dump data filesmetadata`. To enable this set the flag `dump_data = true`.

## templates

No changes.

### CODE REVIEW

**Review by ronpal — DISMISSED** [review]
Love the simplicity of this.

**Review by ronpal — APPROVED** [review]

File: cognite_toolkit/_cdf_tk/commands/dump_data.py [hunk 1]
```diff
@@ -244,6 +253,55 @@ def _get_resource_columns(self) -> list[SchemaColumn]:
...
+            SchemaColumn(name="dataSetExternalId", type="string"),
+            SchemaColumn(name="sourceCreatedTime", type="integer"),
+            SchemaColumn(name="sourceModifiedTime", type="integer"),
+            SchemaColumn(name="securityCategories", type="string", is_array=True),
+            SchemaColumn(name="labels", type="string", is_array=True),
+            SchemaColumn(name="geoLocation", type="json"),
+        ]
+        data_set_ids = self.client.lookup.data_sets.id(self.data_sets) if self.data_sets else []
+        root_ids = self.client.lookup.assets.id(self.hierarchies) if self.hierarchies else []
+        metadata_keys = metadata_key_counts(self.client, "files", data_set_ids or None, root_ids or None)
```
Comment (rsjr) [inline cognite_toolkit/_cdf_tk/commands/dump_data.py:288]: If the result from `metadata_key_counts` is empty, you will get "object is not iterable" error. Should be set to `[]` if that the case (same as you did for `data_set_ids` and `root_ids`

File: cognite_toolkit/_cdf_tk/commands/dump_data.py [hunk 2]
```diff
@@ -244,6 +253,55 @@ def _get_resource_columns(self) -> list[SchemaColumn]:
...
+        return FileMetadataLoader.create_loader(client)
+
+    def _aggregate_count(self, hierarchies: list[str], data_sets: list[str]) -> int:
+        result = self.client.files.aggregate(
+            filter=FileMetadataFilter(
+                data_set_ids=[{"externalId": item} for item in data_sets] or None,
+                asset_subtree_ids=[{"externalId": item} for item in hierarchies] or None,
+            )
+        )
+        return result[0].count if result else 0
```
Comment (rsjr) [inline cognite_toolkit/_cdf_tk/commands/dump_data.py:269]: What's the type for "result" here? Is it FileMetadataList from the SDK?
Wondering what will happen if the indexed result has no "count" field.

File: cognite_toolkit/_cdf_tk/utils/interactive_select.py [hunk 3]
```diff
@@ -115,6 +123,17 @@ def _aggregate_count(self, hierarchies: list[str], data_sets: list[str]) -> int:
...
 
+class FileMetadataInteractiveSelect(AssetCentricInteractiveSelect):
+    def _aggregate_count(self, hierarchies: list[str], data_sets: list[str]) -> int:
+        result = self.client.files.aggregate(
+            filter=FileMetadataFilter(
+                data_set_ids=[{"externalId": item} for item in data_sets] or None,
+                asset_subtree_ids=[{"externalId": item} for item in hierarchies] or None,
+            )
+        )
+        return result[0].count if result else 0
```
Comment (rsjr) [inline cognite_toolkit/_cdf_tk/utils/interactive_select.py:134]: What's the type for "result" here? Is it FileMetadataList from the SDK?
Wondering what will happen if the indexed result has no "count" field.

File: tests/test_unit/test_cdf_tk/test_commands/test_dump_data.py [hunk 4]
```diff
@@ -95,6 +99,69 @@ def test_dump_assets(self, tmp_path: Path) -> None:
...
+                output_dir,
+                clean=True,
+                limit=None,
+                format_="csv",
+                verbose=False,
+            )
+
+        output_csvs = list(output_dir.rglob("*.csv"))
+        assert len(output_csvs) == 1
+        output_csv = output_csvs[0]
```
Comment (rsjr) [inline tests/test_unit/test_cdf_tk/test_commands/test_dump_data.py:145]: Can you have a test case here for output_parquet as well? Looks like it is missing.

File: tests/test_unit/test_cdf_tk/test_commands/test_dump_data.py [hunk 5]
```diff
@@ -95,6 +99,69 @@ def test_dump_assets(self, tmp_path: Path) -> None:
         parquet_files = list(parquet_dir.rglob("*.parquet"))
         assert len(parquet_files) == 1
 
+    def test_dump_filemetadata(self, tmp_path: Path) -> None:
```
Comment (rsjr) [inline tests/test_unit/test_cdf_tk/test_commands/test_dump_data.py:102]: Maybe I couldn't find it, but I would like to see some tests for some non-happy paths edge cases such as:

- Test case with some "bad" files in it (invalid MIME, special chars in name/external Id etc)
- what happens if metadata is empty?

File: tests/test_unit/test_cdf_tk/test_utils/test_interactive_select.py [hunk 6]
```diff
@@ -25,17 +26,47 @@ def select_data_set(choices: list[Choice]) -> list[str]:
...
+        def select_hierarchy(choices: list[Choice]) -> list[str]:
+            assert len(choices) == 2
+            return [choices[1].value]
+
+        answers = ["Data Set", select_data_set, "Hierarchy", select_hierarchy, "Done"]
+        with (
+            monkeypatch_toolkit_client() as client,
+            MockQuestionary(TimeSeriesInteractiveSelect.__module__, monkeypatch, answers),
+        ):
+            client.data_sets.list.return_value = [
```
Comment (rsjr) [inline tests/test_unit/test_cdf_tk/test_utils/test_interactive_select.py:59]: maybe add a test case if data_set is None or []

**Review by rsjr — COMMENTED** [review]
Some comments. Feel free to ping me when you are done so I can re-review it

File: tests/test_unit/test_cdf_tk/test_commands/test_dump_data.py (hunk 4 above)
Comment (doctrino) [reply tests/test_unit/test_cdf_tk/test_commands/test_dump_data.py:145]: That would be testing the TableWrites and not the functionality in this PR. That is tested extensively in https://github.com/cognitedata/toolkit/blob/main/tests/test_unit/test_cdf_tk/test_utils/test_table_writers.py
